engine/crates/engine/registry-v2/src/common/types.rs

```diff
@@ -0,0 +1,207 @@
+                WrappingType::List => wrappers.wrap_list(),
+            })
+    }
+}
+
+pub struct TypeWrappersIter {
+    encoded: u32,
+    mask: u32,
+    next: Option<WrappingType>,
+    last: Option<WrappingType>,
+}
+
+impl Iterator for TypeWrappersIter {
+    type Item = WrappingType;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(next) = self.next.take() {
+            return Some(next);
+        }
+        if (self.mask & NUM_LISTS_MASK) != 0 {
+            if let Some(last) = self.last.take() {
+                return Some(last);
+            }
+            return None;
+        }
+
+        // Otherwise we still have list wrappers
+        let current_is_non_null = (self.encoded & self.mask) != 0;
+        self.mask >>= 1;
+
+        if current_is_non_null {
+            self.next = Some(WrappingType::List);
+            Some(WrappingType::NonNull)
+        } else {
+            Some(WrappingType::List)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{TypeWrappers, WrappingType};
+
+    #[test]
+    fn test_wrappers() {
+        assert_eq!(TypeWrappers::none().iter().collect::<Vec<_>>(), vec![]);
+        assert_eq!(
+            TypeWrappers::none().wrap_non_null().iter().collect::<Vec<_>>(),
+            vec![WrappingType::NonNull]
+        );
+
+        assert_eq!(
+            TypeWrappers::none().wrap_list().iter().collect::<Vec<_>>(),
+            vec![WrappingType::List]
+        );
+
+        assert_eq!(
+            TypeWrappers::none()
+                .wrap_non_null()
+                .wrap_list()
+                .iter()
+                .collect::<Vec<_>>(),
+            vec![WrappingType::List, WrappingType::NonNull]
+        );
+
+        assert_eq!(
+            TypeWrappers::none()
+                .wrap_non_null()
+                .wrap_list()
+                .wrap_non_null()
+                .iter()
+                .collect::<Vec<_>>(),
+            vec![WrappingType::NonNull, WrappingType::List, WrappingType::NonNull]
+        );
+
+        assert_eq!(
+            TypeWrappers::none()
+                .wrap_list()
+                .wrap_list()
+                .wrap_list()
+                .wrap_non_null()
+                .iter()
+                .collect::<Vec<_>>(),
+            vec![
+                WrappingType::NonNull,
+                WrappingType::List,
+                WrappingType::List,
+                WrappingType::List,
+            ]
+        );
+
+        assert_eq!(
+            TypeWrappers::none()
+                .wrap_non_null()
+                .wrap_list()
+                .wrap_non_null()
+                .wrap_list()
+                .iter()
+                .collect::<Vec<_>>(),
+            vec![
+                WrappingType::List,
+                WrappingType::NonNull,
+                WrappingType::List,
+                WrappingType::NonNull
+            ]
+        );
+    }
+}
```

Comment:
Is there a reason to not use the `wrapping` crate we extracted? (https://github.com/grafbase/grafbase/tree/main/engine/crates/wrapping). If there's missing features we could add them there.

Reply:
This module was just copied over wholesale from cynic-parser.  Could almost certainly use that crate instead, though I might hold off that change till I've got everything else in place